rectools/models/vector.py

```diff
@@ -43,6 +46,121 @@ class Factors:
     biases: tp.Optional[np.ndarray] = None
 
 
+class ImplicitRanker:
+    """Ranker for DOT and COSINE similarity distance which uses implicit library matrix factorization topk method"""
+
+    def __init__(self, distance: Distance, subjects_factors: np.ndarray, objects_factors: np.ndarray) -> None:
+        if distance not in (Distance.DOT, Distance.COSINE):
+            raise ValueError(f"ImplicitRanker is not suitable for {distance} distance")
+        self.distance = distance
+        self.subjects_factors = subjects_factors.astype(np.float32)
+        self.objects_factors = objects_factors.astype(np.float32)
+
+        self.subjects_norms: np.ndarray
+        self.objects_norms: np.ndarray
+        if distance == Distance.COSINE:
+            self.subjects_norms = np.linalg.norm(subjects_factors, axis=1).astype(np.float32)
+            self.objects_norms = np.linalg.norm(objects_factors, axis=1).astype(np.float32)
+
+    def _get_neginf_score(self) -> float:
+        dummy_factors = np.array([[1, 2]], dtype=np.float32)
+        neginf = implicit.cpu.topk.topk(  # pylint: disable=c-extension-no-member
+            items=dummy_factors,  # overall whitelist item factors
+            query=dummy_factors,  # query_factors
+            k=1,
+            filter_items=np.array([0]),
+        )[1][0][0]
+        return neginf
+
+    def _get_mask_for_correct_scores(self, scores: np.ndarray) -> tp.List[bool]:
+        num_masked = 0
+        min_score = self._get_neginf_score()
+        for el in np.flip(scores):
+            if el <= min_score:
+                num_masked += 1
+            else:
+                break
+        return [True for _ in range(len(scores) - num_masked)] + [False for _ in range(num_masked)]
+
+    def _process_implicit_scores(
+        self, subject_ids: np.ndarray, ids: np.ndarray, scores: np.ndarray
+    ) -> tp.Tuple[InternalIds, InternalIds, Scores]:
+
+        all_target_ids = []
+        all_reco_ids: tp.List[np.ndarray] = []
+        all_scores: tp.List[np.ndarray] = []
+
+        for i in range(scores.shape[0]):
```

Comment:
<img width="735" alt="image" src="https://github.com/MobileTeleSystems/RecTools/assets/14802567/a11f4c06-48cf-4634-8e8d-b88764fc4ae0">
(Naming is not suggested)

Reply:
fixed